### PR TITLE
Make sure newtypes use linear arrow

### PR DESCRIPTION
**Pull request by monoidal**
Not ready yet, needs the other part of #173.

### CODE REVIEW

File: compiler/rename/RnSource.hs [hunk 1]
```diff
@@ -2038,16 +2038,18 @@ rnConDecl decl@(ConDeclGADT { con_names   = names
...
                   = case args of
                       InfixCon {}  -> pprPanic "rnConDecl" (ppr names)
                       RecCon {}    -> (new_args, new_res_ty)
-                      PrefixCon as | (arg_tys, final_res_ty) <- splitHsFunType new_res_ty
-                                   -> ASSERT( null as )
-                                      -- See Note [GADT abstract syntax] in HsDecls
-                                      (PrefixCon arg_tys, final_res_ty)
+                      PrefixCon as -> let (arg_tys, final_res_ty) = splitHsFunType new_res_ty
+                                          arg_tys' | linearTypes = arg_tys
+                                                   | otherwise   = map (hsLinear . hsThing) arg_tys
```
Comment (aspiwack) [inline compiler/rename/RnSource.hs]: This is elegant.

**Comment by monoidal** [comment]
Ready for review.
The output for T9222 has worsened, but I'd like to fix this separately, outside of this PR.

File: testsuite/tests/polykinds/T9222.stderr [hunk 2]
```diff
@@ -9,7 +9,7 @@ T9222.hs:14:3: error:
       ‘c’ is a rigid type variable bound by
         the type of the constructor ‘Want’:
           forall i1 j1 (a :: (i1, j1)) (b :: i1) (c :: j1).
-          ((a ~ '(b, c)) => Proxy b) -> Want a
+          ((a ~ '(b, c)) => Proxy b) ->@[n] Want a
```
Comment (aspiwack) [inline testsuite/tests/polykinds/T9222.stderr]: Ok to fix in another PR. But do you know what causes this? This is rather puzzling to me, as it doesn't look like what you did should have affected printing (or defaulting).

Comment (monoidal) [reply testsuite/tests/polykinds/T9222.stderr]: Previously, if `-XLinearTypes` was off, GADTs defined using `->` used unrestricted arrow, and it appeared in the error message. After e019469f0e588385cab340163c7e1dcc8980a897, they use the linear arrow. Somewhere we generalize this to multiplicity polymorphism - I didn't check where, but this was already in the code.

Comment (aspiwack) [reply testsuite/tests/polykinds/T9222.stderr]: Interesting. It should have disappeared in the error message. This is an unusual error message, though, so maybe we need to do some magic switching in it.

Note that, also, `n` weirdly appears unbound. This is also interesting.

At any rate: make a task for it. Also, if it helps, this polymorphised type is generated by `dataConUserType` in `DataCon.hs`.

**Pull request merged**